### PR TITLE
modern-cpp homework

### PR DESCRIPTION
**Pull request by jakjan95**
Exercises: 11, 14, 15, 16, 17, 18
Author: @jakjan95

### CODE REVIEW

**Review by MaciejSalamonski — APPROVED** [review]
Good Job @jakjan95

File: shapes/main.cpp [hunk 1]
```diff
@@ -1,80 +1,68 @@
...
+    for (const auto& it : collection)
         if (it)
             it->print();
 }
 
-void printAreas(const Collection& collection)
-{
-    for (const auto & it : collection)
+void printAreas(const Collection& collection) {
+    for (const auto& it : collection)
```
Comment (WBigaj) [inline shapes/main.cpp:40]: I think you should fix formatting by adding bracets to oneliners,since it will make code more readable.

**Review by WBigaj — COMMENTED** [review]
I have checked your code and have one complain.

File: shapes/main.cpp [hunk 2]
```diff
@@ -1,80 +1,68 @@
...
 
-bool areaLessThan10(shared_ptr<Shape> s)
-{
-    if(s)
-        return (s->getArea() < 10);
-    return false;
-}
+auto areaLessThanX = [x = 10](shared_ptr<Shape> s) {if(s){
+    return (s->getArea() < x);} 
+    return false; };
```
Comment (ziobron) [inline shapes/main.cpp]: I don't like this formatting

**Review by ziobron — CHANGES_REQUESTED** [review]
Missing delegating constructors. This feature allows you to call other constructor of your own class on constructors' initalizer list.

**Comment by jakjan95** [comment]
Formatting issue fixed, added delegating constructors instead of constructors with default parameters :)

File: shapes/Square.cpp [hunk 3]
```diff
@@ -7,6 +7,9 @@ Square::Square(Color c)
 Square::Square(double x, Color c)
     : Rectangle(x, x, c) {}
 
+Square::Square(double x)
+    : Rectangle(x, x, Color::Red) {}
```
Comment (ziobron) [inline shapes/Square.cpp:11]: this one is just a base class constructor call. Delegation takes place when you call other constructor of your own class.

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
15 XP za Nowoczesny C++ SFINAE (4447)
5 XP za Nowoczesny C++ alignas (4448)
5 XP za Nowoczesny C++ delegating constructors (4449)
5 XP za Nowoczesny C++ lambda (4450)
5 XP za Nowoczesny C++ structured bindings (4451)
15 XP za Nowoczesny C++ variadic templates (4452)

{ "requirement_ids": [4447, 4448, 4449, 4450, 4451, 4452] }

**Pull request closed without merging**